keps/sig-instrumentation/20191028-metrics-stability-to-beta.md

```diff
@@ -0,0 +1,86 @@
+---
+title: Metrics Stability Framework to Beta
+authors:
+  - "@logicalhan"
+  - "@RainbowMango"
+owning-sig: sig-instrumentation
+participating-sigs:
+  - sig-instrumentation
+reviewers:
+  - "@brancz"
+approvers:
+  - "@brancz"
+editor: "@brancz"
+creation-date: 2019-10-28
+last-updated: 2019-10-28
+status: implementable
+see-also:
+  - 20181106-kubernetes-metrics-overhaul
+  - 20190404-kubernetes-control-plane-metrics-stability
+  - 20190605-metrics-stability-migration
+  - 20190605-metrics-validation-and-verification
+---
+
+# Metrics Stability Framework to Beta
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+
+## Summary
+
+The metrics stability framework has been added to the Kubernetes project as a way to annotate metrics with a supported stability level. Depending on the stability level of a metric, there are some guarantees one can expect as a consumer (i.e. ingester) of a given metric. This document outline required steps to graduate it to Beta.
+
+## Motivation
+
+The metrics stability framework is currently used for defining metrics stability levels for metrics in OSS Kubernetes. The motivation
+of this KEP is to address those feature requests and bug reports to move this feature to the Beta level.
+
+### Goals
+
+These are the planned changes for Beta feature graduation:
+
+* No Kubernetes binaries register metrics to prometheus registries directly.
+* There is a validated import restriction on all kubernetes binaries (except `component-base/metrics`) such that we will fail, in a precommit phase, a direct import of prometheus in kubernetes. This forces all metrics related code to go through the metrics stability framework.
+* All currently deprecated metrics are deprecated using the `DeprecatedVersion` field of metrics options struct.
+* All Kubernetes binaries should have a command flag `--show-hidden-metrics` by which cluster admins can show metrics deprecated in last minor release.
+
+### Non-Goals
+
+These are the issues considered and rejected for Beta:
+
+* Being able to individually turn off a metric (this will be a GA feature).
```

Comment:
turn off or turn on a deprecated metric?

Reply:
No, it can be any metric. Particularly my contention is that a metric should not be considered to be eligible to be promoted to 'stable' *unless* we can actually turn that metric off at runtime through a flag. This will allow us to sidestep problematic metric issues by providing a mechanism for people to turn off a metric, while adhering to the stability guarantees which we are promising. 

If we can't turn off a metric, then we may end up having to break stability guarantees to fix the afflicted codebase (i.e. by removing a metric). This is only a problem for metrics which are marked as stable, since those are the only ones which we are saying have API guarantees.